File: past/rectangleCutting.py
a, b = map(int, input().split(" "),)

memo = [[-1 for _ in range(b+1)] for _ in range(a+1)]

for i in range(min(a, b)+1):
    memo[i][i] = 0

"""
def helper(rows, cols):
    if rows == cols:
        return 0
    elif memo[rows][cols] != -1:
        return memo[rows][cols]
    
    candidateResults = []
    for rowSplit in range(1, (rows // 2)+1):
        lowerHeight = rows - rowSplit
        candidateResults.append(memo[rowSplit][cols] + memo[lowerHeight][cols])
    
    for colSplit in range(1, (cols // 2)+1):
        leftWidth = cols - colSplit
        candidateResults.append(memo[rows][colSplit] + memo[rows][leftWidth])
    
    result = min(candidateResults) + 1

    memo[rows][cols] = result
    if cols <= a and rows <= b:
        memo[cols][rows] = result
    return result
"""

def noneMin(a, b):
    if a == None:
        return b
    else:
        return(min(a,b))

for rows in range(1, a+1):
    for cols in range(1, b+1):
        if rows == cols:
            continue

        result = rows * cols * 2 + 1
        for rowSplit in range(1, (rows // 2)+1):
            lowerHeight = rows - rowSplit
            result = noneMin(result, memo[rowSplit][cols] + memo[lowerHeight][cols])
        
        for colSplit in range(1, (cols // 2)+1):
            leftWidth = cols - colSplit
            result = noneMin(result, memo[rows][colSplit] + memo[rows][leftWidth])

        result += 1
        memo[rows][cols] = result


print(memo[a][b])

File: past/rectangleCutting.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int a, b;
    if (!(cin >> a >> b)) return 0;

    vector<vector<int>> memo(a + 1, vector<int>(b + 1, -1));

    int limit = min(a, b);
    for (int i = 0; i <= limit; ++i) memo[i][i] = 0;

    const int INF = 1e9;

    for (int rows = 1; rows <= a; ++rows) {
        for (int cols = 1; cols <= b; ++cols) {
            if (rows == cols) continue;  

            int result = INF;

            for (int rowSplit = 1; rowSplit <= rows / 2; ++rowSplit) {
                int lowerHeight = rows - rowSplit;
                int cand = memo[rowSplit][cols] + memo[lowerHeight][cols];
                if (cand < result) result = cand;
            }

            for (int colSplit = 1; colSplit <= cols / 2; ++colSplit) {
                int leftWidth = cols - colSplit;
                int cand = memo[rows][colSplit] + memo[rows][leftWidth];
                if (cand < result) result = cand;
            }

            memo[rows][cols] = result + 1;     
        }
    }

    cout << memo[a][b] << '\n';
    return 0;
}
